pennylane_lightning_kokkos/src/simulator/StateVectorKokkos.hpp

```diff
@@ -702,6 +702,187 @@ template <class Precision> class StateVectorKokkos {
+
+    /**
+     * @brief Probabilities for a subset of the full system.
+     *
+     * @param wires Wires will restrict probabilities to a subset
+     * of the full system.
+     * @return Floating point std::vector with probabilities.
+     * The basis columns are rearranged according to wires.
+     */
+    std::vector<Precision> probs(const std::vector<size_t> &wires) {
+        // Determining index that would sort the vector.
+        // This information is needed later.
+        auto sorted_ind_wires = Util::sorting_indices(wires);
+        // Sorting wires.
+        std::vector<size_t> sorted_wires(wires.size());
+
+        for (size_t pos = 0; pos < wires.size(); pos++) {
+            sorted_wires[pos] = wires[sorted_ind_wires[pos]];
+        }
+
+        // Determining probabilities for the sorted wires.
+        const Kokkos::View<Kokkos::complex<Precision> *> arr_data = getData();
+
+        const size_t num_qubits = getNumQubits();
+
+        std::vector<size_t> all_indices =
+            generateBitPatterns(sorted_wires, num_qubits);
+
+        std::vector<size_t> all_offsets = generateBitPatterns(
+            getIndicesAfterExclusion(sorted_wires, num_qubits), num_qubits);
+
+        Kokkos::View<Precision *> d_probabilities("d_probabilities",
+                                                  all_indices.size());
+
+        Kokkos::View<size_t *> d_sorted_ind_wires("d_sorted_ind_wires",
+                                                  sorted_ind_wires.size());
+        Kokkos::View<size_t *> d_all_indices("d_all_indices",
+                                             all_indices.size());
+        Kokkos::View<size_t *> d_all_offsets("d_all_offsets",
+                                             all_offsets.size());
+
+        using UnmanagedSize_tHostView =
+            Kokkos::View<size_t *, Kokkos::HostSpace,
+                         Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
+        Kokkos::deep_copy(
+            d_all_indices,
+            UnmanagedSize_tHostView(all_indices.data(), all_indices.size()));
+        Kokkos::deep_copy(
+            d_all_offsets,
+            UnmanagedSize_tHostView(all_offsets.data(), all_offsets.size()));
+        Kokkos::deep_copy(d_sorted_ind_wires,
+                          UnmanagedSize_tHostView(sorted_ind_wires.data(),
+                                                  sorted_ind_wires.size()));
+
+        /*
+        for (size_t i0 = 0; i0 < all_indices.size(); i0++) {
+            for (size_t j0 = 0; j0 < all_offsets.size(); j0++) {
+                size_t index = all_indices[i0];
+                size_t offset = all_offsets[j0];
+                probabilities[i0] += std::norm(arr_data[index + offset]);
+            }
+        }
+        */
+
+        const size_t N_Ai = all_indices.size();
+        const size_t N_Ao = all_offsets.size();
+
+        Kokkos::parallel_for(
+            Kokkos::RangePolicy<KokkosExecSpace>(0, N_Ai),
+            KOKKOS_LAMBDA(const size_t &i) {
```

Comment:
Possibly use functors here (for cuda compatibility)

Reply:
yes, that's what i'm going to do today.